### PR TITLE
#163 DynamicDependencies: AppLifecycleTests failure in WinRT activation (combase.dll) on >= 20H1 / May 2020 Update / 10.0.19041.0

### PR DESCRIPTION
**Pull request by DrusTheAxe**
Issue [#653](https://github.com/microsoft/ProjectReunion/issues/653) DynamicDependencies: AppLifecycleTests failure in WinRT activation (combase.dll) on >= 20H1 / May 2020 Update / 10.0.19041.0

(All) AppLifecycleTests in user/aeloros/instancing fail activating a WinRT object in `MddBootstrapInitialize` when run on >=20H1.

The failure in combase is related to calls to Detour'd `GetCurrentPackageInfo` and non-Detour'd `GetCurrentPackageInfo3` producing inconsistent results. One is aware a process could have a package graph without package identity (thanks to Detours and DynDep), the other thinks there can't be a package graph without package identity.

The fix is to Detour `GetCurrentPackageInfo3` and implement compatible logic.

Before this change all AppLifecycleTests in user/aeloros/instancing failed
After this change all AppLifecycleTests pass except the File test, but that may be due to a local machine issue (all tests in Main pass).

I wasn't able to reproduce the error with a more controlled test case but the failure to hook and provide `GetCurrentPackageInfo3` is a clear problem that needed solving. This fix unblocks most if not all AppLifecycle development. If the File test is still a problem this at least shrinks the space of the problem to something more bite size (peeling the onion).

### CODE REVIEW

File: test/DynamicDependency/Test_Win32/Test_GetCurrentPackageInfo.cpp [hunk 1]
```diff
@@ -0,0 +1,374 @@
...
+            }
+
+            UINT32 generationId{};
+            UINT32 bufferSize{ static_cast<UINT32>(sizeof(generationId)) };
+            const auto hr{ m_getCurrentPackageInfo3(0, PackageInfoType_PackageInfoGeneration, &bufferSize, &generationId, nullptr) };
+            auto message{ wil::str_printf<wil::unique_process_heap_string>(L"Get...GenerationId: hr:0x%X id:%u\n"
+                                                                           L"          Expected: hr:0x%X id:%u",
+                                                                           hr, generationId, expectedHR, expectedGenerationId) };
+            Assert::IsTrue(true, message.get());
+            OutputDebugStringW(message.get());
```
Comment (DrusTheAxe) [inline test/DynamicDependency/Test_Win32/Test_GetCurrentPackageInfo.cpp:247]: This is to aid debugging and verification. Will be rewritten when we address [Rewrite DynamicDependencies tests using TAEF #559](https://github.com/microsoft/ProjectReunion/issues/559)

**Review by aeloros — APPROVED** [review]

File: dev/Common/AppModel.Identity.h [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#ifndef __APPMODEL_IDENTITY_H
+#define __APPMODEL_IDENTITY_H
+
+#include <appmodel.h>
+
+namespace AppModel::Identity
+{
+inline bool IsPackagedProcess()
```
[the remark this answers is not in the document]
Comment (DrusTheAxe) [reply dev/Common/AppModel.Identity.h:11]: Not that I'm aware of

File: dev/Common/Security.IntegrityLevel.h [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+// Licensed under the MIT License. See LICENSE in the project root for license information.
+
+#ifndef __SECURITY_INTEGRITYLEVEL_H
+#define __SECURITY_INTEGRITYLEVEL_H
+
+namespace Security::IntegrityLevel
+{
+inline DWORD GetIntegrityLevel(HANDLE token = nullptr)
+{
+    wistd::unique_ptr<TOKEN_MANDATORY_LABEL> tokenMandatoryLabel{ wil::get_token_information_failfast<TOKEN_MANDATORY_LABEL>(!token ? GetCurrentThreadEffectiveToken() : token) };
```
[the remark this answers is not in the document]
Comment (DrusTheAxe) [reply dev/Common/Security.IntegrityLevel.h:11]: Not sure if there's any specific line length rule. Overarching guidance is 'good taste' and fair enough regardless the exact rule this is too long. Reformatted for reability.

File: dev/DynamicDependency/MddDetourPackageGraph.cpp [hunk 4]
```diff
@@ -53,12 +55,80 @@ LONG WINAPI DynamicGetCurrentPackageInfo2(
...
+    UINT32* count) = nullptr;
+
+HRESULT WINAPI DynamicGetCurrentPackageInfo3(
+    UINT32 flags,
+    PackageInfoType packageInfoType,
+    UINT32* bufferLength,
+    void* buffer,
+    UINT32* count);
+
+typedef HRESULT (WINAPI* GetCurrentPackageInfo3Function)(
```
[the remark this answers is not in the document]
Comment (DrusTheAxe) [reply dev/DynamicDependency/MddDetourPackageGraph.cpp:79]: I'm not sure if there's a hard rule against one line for >1 parameter, but I'm not a fan :P

I swiped this from UndockedRegFreeWinRT, hence the inconsistency. Reformatted for readability and consistency. Thanks!

(Will get commonalized in a future PR)

**Pull request merged**